File: Data_structures/Binary Tree/avl_tree.cpp
#include<bits/stdc++.h>

using namespace std;

struct Node{
    int key;
    Node* left;
    Node* right;
    int height;
}*root;

Node* insert(Node*, int);
Node* newNode(int);
int height(Node*);
Node* leftRotate(Node*);
Node* rightRotate(Node*);

void preOrder(Node*);

int main()
{
    root = NULL;

    root = insert(root, 10);
    preOrder(root);
    root = insert(root, 20);
    preOrder(root);
    root = insert(root, 30);
    preOrder(root);
    root = insert(root, 40);
    preOrder(root);
    root = insert(root, 50);
    preOrder(root);
    root = insert(root, 25);

    /* The constructed AVL Tree would be  
                30  
            / \  
            20 40  
            / \ \  
        10 25 50  
    */

   printf("preOrder traversal is : ");

   preOrder(root);

    return 0;
}

Node* insert(Node* node, int key)
{
    if(node == NULL)
        return (newNode(key));
    else if(node->key > key)
        node->left = insert(node->left, key);
    else if(node->key < key)
        node->right = insert(node->right, key);
    else
        return node;
    
    node->height = 1 + max(height(node->left), height(node->right));

    int balance = height(node->left) - height(node->right);

    if(balance > 1)
    {
        // Left-Left case
        if (key > node->left->key)
        {
            return rightRotate(node);
        }
        // Left Right case
        else 
        {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }
    }
    else if(balance < -1)
    {
        // Right-Right case
        if (key > node->right->key)
        {
            return leftRotate(node);
        }
        // Right Left case
        else
        {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }
    }

    return node;
}


Node* newNode(int key)
{
    Node* node = new Node();
    node->key = key;
    node->left = NULL;
    node->right = NULL;
    node->height = 1;

    return node;
}

int height(Node* node)
{
    if(node == NULL)
        return 0;

    return node->height;
}

Node* leftRotate(Node* node)
{
    struct Node *temp;
    temp = node->right->left;

    node->right->left = node;
    node->right = temp;

    node->height = 1 + max(height(node->left), height(node->right));
    node->right->height = 1 + max(height(node->right->left), height(node->right->right));

    return node->right;
}

Node* rightRotate(Node* node)
{
    struct Node *temp;

    temp = node->left->right;

    node->left->right = node;
    node->left = temp;

    node->height = 1 + max(height(node->left), height(node->right));
    node->left->height = 1 + max(height(node->left->left), height(node->left->right));

    return node->left;
}

void preOrder(Node * root)
{
    if (root != NULL)
    {
        printf("%d ", root->key);
        preOrder(root->left);
        preOrder(root->right);
    }
}

File: Data_structures/AVL Tree/AVLTree.java
public class AVLTree {
    private int height, key;
    private AVLTree left, right;
    private static final AVLTree NIL = new AVLTree();

    public AVLTree(int key) {
        this.key = key;
        left = right = NIL;
    }

    public AVLTree(int key, AVLTree left, AVLTree right) {
        this.key = key;
        this.left = left;
        this.right = right;
        height = 1 + Math.max(left.height, right.height);
    }

    public AVLTree() {
        left = right = this;
        height = -1;
    }

    public boolean add(int key) {
        int oldSize = size();
        grow(key);
        return size() > oldSize;
    }

    private AVLTree grow(int key) {
        if (this == NIL)
            return new AVLTree(key);
        if (this.key == key)
            return this;// avoid key duplicacy
        if (key < this.key)
            left = left.grow(key);
        else
            right = right.grow(key);
        rebalance();
        height = 1 + Math.max(left.height, right.height);
        return this;
    }

    private void rebalance() {
        if (left.height > right.height + 1) {
            if (left.right.height > left.left.height)
                left.leftRotate();
            rightRotate();
        } else if (right.height > left.height + 1) {
            if (right.left.height > right.right.height)
                right.rightRotate();
            leftRotate();
        }
    }

    private void leftRotate() {
        left = new AVLTree(key, left, right.left);
        key = right.key;
        right = right.right;
    }

    private void rightRotate() {
        right = new AVLTree(key, left.right, right);
        key = left.key;
        left = left.left;
    }

    public int size() {
        if (this == NIL || height == -1)
            return 0;
        return 1 + left.size() + right.size();
    }

    @Override
    public String toString() {
        if (this == NIL)
            return "";
        return "" + left + " " + key + " " + right;
    }

    public static void main(String[] args) {
        AVLTree tree=new AVLTree(4);
    	tree.add(3);
    	tree.add(13);
    	tree.add(11);
    	tree.add(9);
    	tree.add(45);
    	tree.add(5);
    	System.out.println(tree.height);
    	System.out.println(tree);
    }

}